src/NativeScripts.purs

```diff
@@ -0,0 +1,91 @@
+module NativeScripts
+  ( NativeScriptHash(NativeScriptHash)
+  , nativeScriptHash
+  , getMaximumSigners
+  ) where
+
+import Prelude
+
+import Cardano.Types.Transaction
+  ( NativeScript
+      ( ScriptPubkey
+      , ScriptAll
+      , ScriptAny
+      , ScriptNOfK
+      , TimelockStart
+      , TimelockExpiry
+      )
+  )
+import Data.Array as Array
+import Data.Foldable (foldr, maximumBy)
+import Data.Function (on)
+import Data.List (List(Cons, Nil), (:))
+import Data.List as List
+import Data.Maybe (Maybe, fromMaybe)
+import Data.Newtype (class Newtype, wrap)
+import Data.Set (Set)
+import Data.Set as Set
+import Serialization.Hash (Ed25519KeyHash, ScriptHash)
+import Serialization.Hash as Hashing
+import Serialization.NativeScript (convertNativeScript)
+
+newtype NativeScriptHash = NativeScriptHash ScriptHash
+
+derive instance Newtype NativeScriptHash _
+derive newtype instance Eq NativeScriptHash
+derive newtype instance Ord NativeScriptHash
+
+instance Show NativeScriptHash where
+  show (NativeScriptHash sh) = "(NativeScriptHash " <> show sh <> ")"
+
+nativeScriptHash :: NativeScript -> Maybe NativeScriptHash
+nativeScriptHash ns = wrap <<< Hashing.nativeScriptHash <$> convertNativeScript
+  ns
+
+-- | `SetChoice` is an internal type representing internal state of
+-- | `getMaximumSigners` algorithm.
+type SetChoice a = Array (Set a)
+
+anyChoice :: forall a. Ord a => SetChoice a -> SetChoice a -> SetChoice a
+anyChoice as bs = Array.nub $ as <> bs
+
+allChoice :: forall a. Ord a => SetChoice a -> SetChoice a -> SetChoice a
+allChoice as bs = (<>) <$> as <*> bs
+
+subsetsOfLength :: forall a. Int -> Array a -> Array (Array a)
+subsetsOfLength n =
+  List.fromFoldable >>> sublists n >>> List.toUnfoldable >>> map
+    List.toUnfoldable
+
+sublists :: forall a. Int -> List a -> List (List a)
+sublists n xs = List.take (List.length xs - n + 1) $ sublists' n xs
+  where
+  sublists' _ Nil = Cons Nil Nil
```

Comment:
I know it's obvious, but for the same of consistency please add a type signature